tools/asynccmd/range.go

```diff
@@ -0,0 +1,83 @@
+package asynccmd
+
+import (
+	"bytes"
+	"sort"
+)
+
+type matchRange struct {
+	first int
+	last  int
+}
+
+var emptyRange = matchRange{}
+
+// allRanges returns every indexes of instance of pattern in b, or nil if pattern is not present in b.
+func allRanges(b, pattern []byte) (ranges []matchRange) {
+	i := 0
+	for {
+		sub := b[i:len(b)]
+		idx := bytes.Index(sub, pattern)
+		if idx == -1 {
+			return
+		}
+
+		ranges = append(ranges, matchRange{first: idx + i, last: idx + i + len(pattern)})
+		i += idx + 1
+	}
+}
+
+// isOverlapping returns true if base and compare ranges overlaps each other, otherwise false.
+func isOverlapping(base, compare matchRange) bool {
+	return compare.first >= base.first && compare.first <= base.last
+}
+
+// mergeRanges returns the union of r1 and r2, r1 and r2 have to be overlapping ranges.
+func mergeRanges(r1, r2 matchRange) matchRange {
+	first := r1.first
+	if r2.first < first {
+		first = r2.first
+	}
+	last := r1.last
+	if r2.last > last {
+		last = r2.last
+	}
+	return matchRange{first: first, last: last}
+}
+
+// mergeAllRanges merges every overlapping ranges in r.
+func mergeAllRanges(r []matchRange) []matchRange {
+	ranges := append([]matchRange{}, r...)
```

Comment:
no need for `ranges` variable, use just `r`. I would rename r to `ranges` or `rs`